python/src/aac/lang/lsp/language_server.py

```diff
@@ -45,16 +51,53 @@ def configure_lsp(self):
         self.code_completion_provider = self.code_completion_provider or CodeCompletionProvider()
         self.setup_features()
 
-        logging.debug("AaC LSP initialized.")
-
     def setup_features(self) -> None:
         """Configure the server with the supported features."""
         server = self.language_server
+        server.sync_kind = TextDocumentSyncKind.FULL
+
+        @server.feature(methods.TEXT_DOCUMENT_DID_OPEN)
+        async def did_open(ls, params: DidOpenTextDocumentParams):
+            """Text document did open notification."""
+            logging.info(f"Text document opened by LSP client {params.text_document.uri}.")
+
+            file_uri = params.text_document.uri
+            managed_file = self.workspace_files.get(file_uri)
+
+            if not managed_file:
+                managed_file = ManagedWorkspaceFile(file_uri)
+                self.workspace_files[file_uri] = managed_file
+
+            managed_file.is_client_managed = True
+            _, file_path = file_uri.split("file://")
+            self.language_context.add_definitions_to_context(parse(file_path))
+
+        @server.feature(methods.TEXT_DOCUMENT_DID_CLOSE)
+        def did_close(server: LanguageServer, params: DidCloseTextDocumentParams):
+            """Text document did close notification."""
+            logging.info(f"Text document closed by LSP client {params.text_document.uri}.")
+
+            file_uri = params.text_document.uri
+            managed_file = self.workspace_files.get(file_uri)
+            managed_file.is_client_managed = False
+
+        @server.feature(methods.TEXT_DOCUMENT_DID_CHANGE)
```

Comment:
We may want to change this to `TEXT_DOCUMENT_DID_SAVE` as `DID_CHANGE` fires when there are unsaved changes. Changing to an on-save model will be a little less resource intensive, and the LSP won't be preemptively complaining about errors/warnings while the user is actively typing.

Reply:
> Reading the spec, it seems like it may be better to use the [textDocument/willSave](https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#textDocument_willSave) or textDocument/willSaveWaitUntil notifications since they would allow us to check the contents before writing to disk.

In terms of updating the LanguageContext, I don't see extra value in leveraging pre-save events since the LSP won't do anything to alter what's being written to disk. 

I've captured your comment and addressing it in this issue: https://github.com/jondavid-black/AaC/issues/331